Translate Hooks Overview (#80)
reactjs/pt-BR.reactjs.org: content/docs/hooks-overview.md

---
id: hooks-overview
title: Hooks à primeira vista
permalink: docs/hooks-overview.html
next: hooks-state.html
prev: hooks-intro.html
---

> *Hooks* são uma nova adição no React 16.8. Eles permitem que você use o estado e outros recursos do React sem escrever uma classe.

Reviewer: ```suggestion
*Hooks* são uma nova adição no React 16.8. Eles permitem que você use state  e outros recursos do React sem escrever uma classe.
```

"State" como _feature_  não precisa ser traduzido, ok?

Author: Tem razão, acabei traduzindo state em vários outros lugares também sem querer, corrigi em todos, obrigado!

> Hooks são [compativeis com versões anteriores](/docs/hooks-intro.html#no-breaking-changes). Esta página fornece uma visão geral de Hooks para usuários experientes em React. Esta é uma visão geral rápida. Se você se sentir confuso, procure uma caixa amarela com esta:

Reviewer: ```suggestion
Hooks são [retrocompatíveis](/docs/hooks-intro.html#no-breaking-changes). Esta página fornece uma visão geral de Hooks para usuários experientes em React. Esta é uma visão geral rápida. Se você se sentir confuso, procure uma caixa amarela com esta:
```

Reviewer 2: faltou um `como` na ultima frase:
```suggestion
Hooks são [retrocompatíveis](/docs/hooks-intro.html#no-breaking-changes). Esta página fornece uma visão geral de Hooks para usuários experientes em React. Esta é uma visão geral rápida. Se você se sentir confuso, procure uma caixa amarela como esta:
```

Esta outra frase aqui `Esta é uma visão geral rápida.` parece um pouco repetitiva, mas esta igual o original hehe

>Explicação Detalhada
>

> >Leia a [Motivação](/docs/hooks-intro.html#motivation) para entender porque estamos introruzindo Hooks para o React.

Reviewer: ```suggestion
>Leia a [Motivação](/docs/hooks-intro.html#motivation) para entender porque estamos introduzindo Hooks para o React.
```

O Único argumento para `useState` é o estado inicial. No exemplo acima, é `0` porque nosso contador começa do zero. Perceba que diferente de `this.state`, o estado não precisa ser um objeto -- apesar de que possa ser se você quiser. O argumento de estado inicial é utilizado apenas durante a primeira renderização.

#### Declarando multiplas variáveis de estado {#declaring-multiple-state-variables}

Você pode utilizar o State Hook mais de uma vez em um único componente:

```js
function ExampleWithManyStates() {
  // Declara várias variáveis de estado!
  const [idade, alterarIdade] = useState(42);
  const [fruta, alterarFruta] = useState('banana');

>   const [tarefas, alterarTarefas] = useState([{ text: 'Aprender Hooks' }]);

Author: Como não vamos traduzir as variáveis e funções, acho que fica estranho traduzir o texto, vou deixar "Learn" mesmo

  // ...
```

Hooks permitem a você organizar efeitos colaterais *(side effects)* em um componente por partes relacionadas (como adicionar e remover uma subscrição), em vez de forçar uma divisão baseada nos métodos de ciclo de vida.

> Explicação Detalhada
>
> Você pode aprender mais sobre `useEffect` na sua página dedicada: [Utilizando o Hook de Efeito](/docs/hooks-effect.html).

## Regras dos Hooks

Hooks são funções JavaScript, mas eles impões duas regras adicionais:

* Apenas chame Hooks **no nível mais alto**. Não chame Hooks dentro de loops, condições ou funções aninhadas.

> * Apenas chame Hooks **de componentes funcionais**. Não chame Hooks de funções JavaScript comuns. (Há apenas um outro lugar válido para se chamar Hooks -- dentro dos seus Hooks customizados. Iremos aprender sobre eles em breve.)

Reviewer: ```suggestion
* Apenas chame Hooks **de componentes funcionais**. Não chame Hooks de funções JavaScript comuns. (Há apenas um outro lugar válido para se chamar Hooks -- dentro dos seus próprios Hooks customizados. Iremos aprender sobre eles em breve.)
```

> Nós fornecemos um [plugin de linter](https://www.npmjs.com/package/eslint-plugin-react-hooks) para assegurar essas regras automáticamente. Entendemos que essas regras podem parecer limitantes ou confusas a princípio, mas são essenciais para fazer com que os Hooks funcionem bem.

Reviewer: ```suggestion
Nós fornecemos um [plugin de linter](https://www.npmjs.com/package/eslint-plugin-react-hooks) para assegurar essas regras automaticamente. Entendemos que essas regras podem parecer limitantes ou confusas a princípio, mas são essenciais para fazer com que os Hooks funcionem bem.
```

> Explicação Detalhada
>
> Você pode aprender mais sobre essas regras na sua página dedicada: [Regras dos Hooks](/docs/hooks-rules.html).

> ## Construindo Seus Próprios Hooks

Reviewer: ```suggestion
## 💡 Construindo Seus Próprios Hooks {#-building-your-own-hooks}
```

Author: Acho que isso de remover os links e alterações no código foram feitas pelo crowdin sem querer :/

> Às vezes, queremos reutilizar algumas lógicas de estado entre componentes. Tradicionalmente, haviam duas soluções populares para este problema: [componentes de ordem superior](/docs/higher-order-components.html) e [renderização de props](/docs/render-props.html). Hooks Customizados te permitem fazer isso, mas sem adicionar mais componentes para a sua árvore.

Reviewer: ```suggestion
Às vezes, queremos reutilizar algumas lógicas de estado entre componentes. Tradicionalmente, haviam duas soluções populares para este problema: [componentes de ordem superior (HOCs)](/docs/higher-order-components.html) e [renderização de props](/docs/render-props.html). Hooks Customizados te permitem fazer isso, mas sem adicionar mais componentes para a sua árvore.
```

Nós colocamos "stateful logic" como "lógica com estados" no Glossário, mas aqui, "de estado" me parece soar melhor mesmo.

Estou em dúvida se devemos revisar o termo. Gostaria de uma segunda opinão, o que acham?

> Anteriormente nesta página, nós introduzimos um componente `FriendStatus` que chama os Hooks `useState` e `useEffect` para subscrever-se ao status de online de um amigo. Digamos que também precisaremos reutilizar essa subscrição lógica em outro componente.

Reviewer: ```suggestion
Anteriormente nesta página, nós introduzimos um componente `FriendStatus` que chama os Hooks `useState` e `useEffect` para subscrever-se ao status de online de um amigo. Digamos que também precisaremos reutilizar essa lógica de subscrição em outro componente.
```

Primeiramente, iremos extrair esta lógica para um Hook customizado chamado `useFriendStatus`:

```js{3}

> import { useState, useEffect } from 'react';

Reviewer: ```suggestion
import React, { useState, useEffect } from 'react';
```

}
```

Que recebe `friendID` como um argumento, e sempre retorna se nosso amigo está online.

> Agora podemos utilizar dos dois componentes:

Reviewer: ```suggestion
Agora podemos utilizar a partir dos dois componentes:
```

}
```

> O estato destes componentes é completamente independente. Hooks são a forma de reutilizar *lógica de estado*, não o estado em si. De fato, cada *chamada* para um Hook tem um estado completamente isolado -- então você pode até utilizar o mesmo Hook custom duas vezes em um componente.

Reviewer: ```suggestion
O estado (state) destes componentes é completamente independente. Hooks são a forma de reutilizar *lógica de estado*, não o estado em si. De fato, cada *chamada* para um Hook tem um estado completamente isolado -- então você pode até utilizar o mesmo Hook custom duas vezes em um componente.
```

Hooks customizados são mais uma convenção do que uma funcionalidade. Se o nome de uma função começa com "`use`" e chama outros Hooks, consideramos que é um Hoook customizado. A convenção de nome `useSomething` é como nosso plugin de linter é capaz de encontrar bugs no nosso código que utiliza Hooks.

> Você pode escrever Hooks customizados que abrangem uma ampla gama de casos de uso, como manipulação de formulários, animações, subscrições declarativas, temporizadores e provavelmente muitos outros que não consideramos. Estamos animados para ver quais Hooks customizados a comunidade React irá criar.

Reviewer: Essa linha ficou ótima! Parabéns! 👍

Author: Obrigado! :rocket:

> Explicação Detalhada
>
> Você pode aprender mais sobre Hooks Customizados na sua página dedicada: [Construindo Seus Próprios Hooks](/docs/hooks-custom.html).

> ## Outros Hooks

Reviewer: ```suggestion
## 🔌 Outros Hooks {#-other-hooks}
```

}
```

E [`useReducer`](/docs/hooks-reference.html#usereducer) permite gerenciar state local de componentes complexos com um reducer:

```js{2}
function Todos() {
  const [todos, dispatch] = useReducer(todosReducer);
  // ...
```

> Explicação Detalhada
>
> Você pode aprender mais sobre todos os Hooks internos na sua página dedicada: [Referência da API de Hooks](/docs/hooks-reference.html).

> ## Próximos Passos

Reviewer: ```suggestion
## Próximos Passos {#next-steps}
```

Reviewer: @gabrielbs Essa aqui tb faltou, hehehe, esse GitHub escondendo as coisas!

> Ufa, essa foi rápido! Se algumas coisas não fizeram sentido ou se você quiser aprender mais detalhes, você pode ler as próximas páginas, começando com a documentação de [State Hook](/docs/hooks-state.html).

Reviewer: ```suggestion
Ufa, essa foi rápida! Se algumas coisas não fizeram sentido ou se você quiser aprender mais detalhes, você pode ler as próximas páginas, começando com a documentação de [State Hook](/docs/hooks-state.html).
```

Você também pode conferir a [Referência de API de Hooks ](/docs/hooks-reference.html) e o [FAQ de Hooks](/docs/hooks-faq.html).

> Por fim, não perca a [página de introdução](/docs/hooks-intro.html) que explica *o porquê* estamos adicionando Hooks e como iremos começar a utilizá-los lado a lado com as classes -- sem reescrever os nossos apps.

Reviewer: ```suggestion
Por fim, não perca a [página de introdução](/docs/hooks-intro.html) que explica *porque* estamos adicionando Hooks e como iremos começar a utilizá-los lado a lado com as classes -- sem reescrever os nossos apps.
```
